### PR TITLE
ND-49 : Odoberanie sprav v strede navodu v2

### PR DESCRIPTION
**Pull request by smith558**
Funkcionalita pre admina vkladat notification subscription form pre rozne subscription type na ním specifikovane miesto v navode (**journeys** aj **steps**) pomocou tagu

`<notification-subscription types="*_SUBSCRIPTION_TYPES_" />`

image

image

**SYNTAX:**

```html
<notification-subscription types="BlankJourneySubscription" />
```

Je mozne pridat niekolko typov naraz

```html
<notification-subscription types="BlankJourneySubscription,NextVoteSubscription" />
```

Vklada component do
```html
<div types="BlankJourneySubscription" aria-live="polite" class="notification-subscription">...</div>
```

Je mozne doplnat dalsie atributy, napr.
```html
<notification-subscription types="BlankJourneySubscription" style="height: 300px" />
```
**TODO:**
- [x] Functionality
- [x] Review
- [x] RSpec

### CODE REVIEW

File: spec/features/embedded_apps_spec.rb [hunk 1]
```diff
@@ -1,11 +1,35 @@
...
+    let!(:journey) { create(:journey) }
+    let!(:child_birth_step) { create(:step, journey: journey, description: '<embedded-app app-id="narodenie-rodny-list" />') }
+
+    scenario 'inserts the Child Birth Picking up Protocol app' do
+      visit journey_step_path(journey, child_birth_step)
+      expect(page).to have_content('Slobodná')
+    end
+  end
+
+  feature 'Notification subscriptions' do
```
Comment (eronisko) [inline spec/features/embedded_apps_spec.rb]: Tieto testy by mali ist do zvlast suboru, e.g. `spec/features/notification_subscription_component_spec.rb` alebo `spec/helpers/custom_components_helper_spec.rb`

**Comment by smith558** [comment]
@eronisko takze kompletne oddelit custom components (napr notification-subscription component) od embedded apps? Teraz je to pomiesane troska lebo aj ked pozres https://github.com/slovensko-digital/navody.digital/blob/master/spec/helpers/custom_components_helper_spec.rb tak tam to je pripravene na testovanie spolu

File: spec/helpers/custom_components_helper_spec.rb [hunk 2]
```diff
@@ -1,6 +1,6 @@
 require 'rails_helper'
 
-RSpec.describe NotificationSubscriptionsHelper, type: :helper do
+RSpec.describe CustomComponentsHelper, type: :helper do
```
Comment (eronisko) [inline spec/helpers/custom_components_helper_spec.rb]: Dobry postreh 👍

File: spec/helpers/custom_components_helper_spec.rb [hunk 3]
```diff
@@ -27,5 +27,32 @@
         expect(result).to include "Slobodná"
       end
     end
+
+    describe '<notification-subscription />' do
+      before(:each) do
+        create_fake_user
+      end
+
+      it 'renders an embedded notification subscription component' do
```
Comment (eronisko) [inline spec/helpers/custom_components_helper_spec.rb]: bez `embedded`

File: spec/helpers/custom_components_helper_spec.rb [hunk 4]
```diff
@@ -27,5 +27,32 @@
...
+        result = helper.raw_with_custom_components('<notification-subscription type="BlankJourneySubscription" /><notification-subscription type="BlankJourneySubscription" />')
+        expect(Nokogiri(result).css('input[name="notification_subscription_group[subscriptions][]"]').size).to eq 2
+      end
+
+      it 'supports component being deeper' do
+        result = helper.raw_with_custom_components('<div><notification-subscription type="BlankJourneySubscription" /></div>')
+        expect(result).to include 'Zašleme Vám e-mail, keď vytvoríme tento návod alebo sa bude diať niečo relevantné.'
+      end
+
+      def create_fake_user
```
Comment (eronisko) [inline spec/helpers/custom_components_helper_spec.rb]: Nestaci toto strcit do `before(:each)` ?

File: spec/helpers/custom_components_helper_spec.rb [hunk 5]
```diff
@@ -27,5 +27,32 @@
...
+        create_fake_user
+      end
+
+      it 'renders an embedded notification subscription component' do
+        result = helper.raw_with_custom_components('<notification-subscription type="BlankJourneySubscription" />')
+
+        expect(result).to include 'Zašleme Vám e-mail, keď vytvoríme tento návod alebo sa bude diať niečo relevantné.'
+      end
+
+      it 'supports multiple occurences' do
```
Comment (eronisko) [inline spec/helpers/custom_components_helper_spec.rb]: Chyba mi tu este test na viac typov, e.g. 
```html
<notification-subscription types="BlankJourneySubscription,SomeOtherSubscription" />
```

Comment (smith558) [reply spec/helpers/custom_components_helper_spec.rb]: Uz je v feature testoch, nechcel som zbytocne duplikovat, ci treba?

Comment (eronisko) [reply spec/helpers/custom_components_helper_spec.rb]: Mas pravdu, prehod ten `inserts multiple embedded notification subscription components` z feature sem do helper specov: `supports multiple 'types' values`

Vo feature testoch testujeme ci cely stack funguje, tak ako chceme. Taketo 'full-stack' testy su super, ale byvaju spravidla pomale, takze ja v nich radsej riesim len high-level funkcionalitu pripadne uzivatelske interakcie.

Jednotlive edge casey potom spravilda testujeme v takychto 'unit' specoch, pretoze su omnoho rychlejsie (netestujes cely stack, iba tie jednotlive view rendre).

File: spec/features/custom_components/notification_subscription_component_spec.rb [hunk 6]
```diff
@@ -0,0 +1,22 @@
+require 'rails_helper'
+
+RSpec.feature "Notification subscription components" do
+  let!(:journey) { create(:journey, description: '<notification-subscription type="BlankJourneySubscription" />') }
+  let!(:step) { create(:step, journey: journey, description: '<notification-subscription type="BlankJourneySubscription, NextVoteSubscription" />') }
+
+  scenario 'inserts embedded notification subscription component in a journey' do
```
Comment (eronisko) [inline spec/features/custom_components/notification_subscription_component_spec.rb]: Este odtialto prosim povyhadzuj tie `embedded ` slova

Comment (smith558) [reply spec/features/custom_components/notification_subscription_component_spec.rb]: Akurat som pushoval update, opravil som aj ten `cin `literal

File: app/helpers/custom_components_helper.rb [hunk 7]
```diff
@@ -19,4 +23,9 @@ def render_embedded_app(fragment)
       return render template: 'apps/child_birth_app/picking_up_protocol/start', layout: 'layouts/embedded_app'
     end
   end
+
+  def render_notification_subscription(fragment)
+    @extra_attributes = fragment.attributes.map { |k,v| [k.to_sym, v.value] }.to_h
+    return render_notification_subscription_component(@extra_attributes[:type].split(%r{,\s*}))
```
[the remark this answers is not in the document]
Comment (smith558) [reply app/helpers/custom_components_helper.rb]: @eronisko viem to spravit aj s `@extra_attributes`, ide o to ci to treba k tomuto komponentu, ta methoda `render_notification_subscription_component(subscription_types, journey: nil)` tam ma v svojich views velmi vela stylov uz, to by sa bylo. Ale ak len kvoli `class `a `id `a inym tagom tak mozem to pushnut hned

[the remark this answers is not in the document]
Comment (eronisko) [reply app/helpers/custom_components_helper.rb]: Dajme zatial bez `extra_attrs` ak je to ostara. Vzdy sa to da doplnit. Tak nateraz vyhryzni len `types`

File: app/helpers/custom_components_helper.rb [hunk 8]
```diff
@@ -19,4 +23,10 @@ def render_embedded_app(fragment)
       return render template: 'apps/child_birth_app/picking_up_protocol/start', layout: 'layouts/embedded_app'
     end
   end
+
+  def render_notification_subscription(fragment)
+    @types = fragment[:types].split(',')
```
Comment (eronisko) [inline app/helpers/custom_components_helper.rb]: `@types` staci ako `types` (neptorebujeme to nikde inde)

File: app/helpers/custom_components_helper.rb [hunk 9]
```diff
@@ -19,4 +23,10 @@ def render_embedded_app(fragment)
...
+    @extra_attributes = fragment.attributes.except('types').map { |k,v| [k.to_sym, v.value] }.to_h
```
Comment (eronisko) [inline app/helpers/custom_components_helper.rb]: Toto teda mozeme dat von.

File: spec/features/embedded_apps_spec.rb [hunk 10]
```diff
@@ -1,10 +1,10 @@
...
 
-RSpec.feature "Embedded apps", type: :feature do
-  let!(:journey) {create(:journey)}
-  let!(:child_birth_step) {create(:step, journey: journey, description: '<embedded-app app-id="narodenie-rodny-list" />')}
+RSpec.feature "Embedded apps" do
+  let!(:journey) { create(:journey) }
+  let!(:child_birth_step) { create(:step, journey: journey, description: '<embedded-app app-id="narodenie-rodny-list" />') }
 
-  scenario 'As a user I want to see an the Child Birth Picking up Protocol app' do
+  scenario 'inserts the Child Birth Picking up Protocol app' do
```
Comment (eronisko) [inline spec/features/embedded_apps_spec.rb]: Toto prosim nechaj ako to bolo. Je to take ukecane, ale je to [user story](https://www.mountaingoatsoftware.com/agile/user-stories).

File: app/helpers/custom_components_helper.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (smith558) [reply app/helpers/custom_components_helper.rb]: Pushol som ako by to vyzeralo s `@extra_attributes`. Povedz mi co si o tom myslis, mozeme revertnut naspät vzdy🤔

File: app/helpers/custom_components_helper.rb (hunk 9 above)
Comment (eronisko) [reply app/helpers/custom_components_helper.rb]: Ak si sa `@extra_attributes` rozhodol ponechat, sprav z toho a z `@types` prosim local variables, t.j. `extra_attributes` a `types`

File: app/views/notification_subscriptions/_layout.js.erb [hunk 11]
```diff
@@ -0,0 +1 @@
+$("div[types='<%= @types&.join(",") %>']").html("<%= escape_javascript(yield) %>");
```
Comment (eronisko) [inline app/views/notification_subscriptions/_layout.js.erb]: Naco ti je tu JS layout? Robis dake ajaxove operacie?

File: app/views/notification_subscriptions/_layout.html.erb [hunk 12]
```diff
@@ -0,0 +1,3 @@
+<%= tag.div 'types': @types&.join(","), 'aria-live': 'polite', class: 'notification-subscription', **@extra_attributes || Hash.new do %>
```
Comment (eronisko) [inline app/views/notification_subscriptions/_layout.html.erb]: Naozaj tu potrebujeme aj partial aj layout?

File: spec/features/embedded_apps_spec.rb [hunk 13]
```diff
@@ -1,6 +1,6 @@
 require 'rails_helper'
 
-RSpec.feature "Embedded apps", type: :feature do
+RSpec.feature "Embedded apps" do
```
Comment (eronisko) [inline spec/features/embedded_apps_spec.rb]: Preco si tu zmazal `type: :feature`? Ma to daky dovod?

Comment (smith558) [reply spec/features/embedded_apps_spec.rb]: Nedeklarujeme to uz vlastne tu '_RSpec.**feature** "Embedded apps", type: :feature do_' , my to pride ako duplikacia.

feature test je vlastne syntaxne
_RSpec.describe "Embedded apps", type: :feature do_
aj
 _RSpec.feature "Embedded apps" do_
to iste ci?

File: app/helpers/custom_components_helper.rb (hunk 9 above)
Comment (smith558) [reply app/helpers/custom_components_helper.rb]: Ako ich potom pouzijem vo view mimo helpera?

File: app/views/notification_subscriptions/_layout.html.erb (hunk 12 above)
Comment (smith558) [reply app/views/notification_subscriptions/_layout.html.erb]: Nakoniec nie

File: app/helpers/custom_components_helper.rb (hunk 9 above)
Comment (eronisko) [reply app/helpers/custom_components_helper.rb]: Vies pouzit toto? https://guides.rubyonrails.org/layouts_and_rendering.html#passing-local-variables

Comment (smith558) [reply app/helpers/custom_components_helper.rb]: To je super feature, o tom som nevedel

File: app/helpers/notification_subscriptions_helper.rb [hunk 14]
```diff
@@ -1,8 +1,8 @@
 module NotificationSubscriptionsHelper
-  def render_notification_subscription_component(subscription_types, journey: nil)
+  def render_notification_subscription_component(subscription_types, extra_attributes = nil, journey = nil)
```
Comment (eronisko) [inline app/helpers/notification_subscriptions_helper.rb]: Veci ako `extra_attributes` byvaju zvacsa ako posledny parameter s defaultom `{}`, t.j.
```suggestion
  def render_notification_subscription_component(subscription_types, journey = nil, extra_attributes = {})
```

Pozri napr https://apidock.com/rails/ActionView/Helpers/FormTagHelper/form_tag (najdi tam 'Show source')

File: app/helpers/notification_subscriptions_helper.rb [hunk 15]
```diff
@@ -1,8 +1,8 @@
...
     group = NotificationSubscriptionGroup.new(subscription_types: subscription_types)
     group.user = current_user
     group.journey = journey
-    render partial: 'notification_subscriptions/form', object: group
+    render partial: 'notification_subscriptions/form', object: group, locals: {extra_attributes: extra_attributes}
```
Comment (eronisko) [inline app/helpers/notification_subscriptions_helper.rb]:
```suggestion
    render partial: 'notification_subscriptions/form', object: group, locals: { extra_attributes: extra_attributes }
```

File: app/helpers/notification_subscriptions_helper.rb (hunk 14 above)
Comment (smith558) [reply app/helpers/notification_subscriptions_helper.rb]: Ja uz tu defaultnu hodnotu riesim vlastne priamo v partiali s ****extra_attributes || Hash.new**, take riesenie sa ti asi nepaci?
`<%= tag.div 'types': form.subscription_types&.join(","), 'aria-live': 'polite', class: 'notification-subscription', **extra_attributes || Hash.new do %>`

Comment (smith558) [reply app/helpers/notification_subscriptions_helper.rb]: na urovni helpera je to lepsie, implementujem

File: app/helpers/notification_subscriptions_helper.rb [hunk 16]
```diff
@@ -1,8 +1,8 @@
 module NotificationSubscriptionsHelper
-  def render_notification_subscription_component(subscription_types, journey: nil)
+  def render_notification_subscription_component(subscription_types: {}, journey: nil, extra_attributes: {})
```
Comment (eronisko) [inline app/helpers/notification_subscriptions_helper.rb]: Taketo keyword argumenty su v Ruby/Rails nestandardne (co je v Rails problematicke, kedze tu razime convention over configuration). Navrhujem revertnut https://github.com/slovensko-digital/navody.digital/pull/255/commits/415eb2a068c5bf1c5a38c19d2f191f9f6e20b7c7 a spravit tak ako som pisal [vyssie](https://github.com/slovensko-digital/navody.digital/pull/255#discussion_r322025490):
```suggestion
  def render_notification_subscription_component(subscription_types, journey = nil, extra_attributes = {})
```

File: app/helpers/notification_subscriptions_helper.rb (hunk 14 above)
Comment (eronisko) [reply app/helpers/notification_subscriptions_helper.rb]: Hej, takato logika (`**extra_attributes || Hash.new`) do views nepatri, v helpri je to lepsie.

File: app/helpers/notification_subscriptions_helper.rb (hunk 16 above)
Comment (smith558) [reply app/helpers/notification_subscriptions_helper.rb]: @eronisko  tam mam ale problem ze ak nechcem zakazdim zadavat `journey: nil` argument tak mi bude optional `journey` brat mylne `extra_attributes` v invokacii `render_notification_subscription_component(subscription_types, extra_attributes)`

definicia `def render_notification_subscription_component(subscription_types, extra_attributes = {},  journey = nil)` moze byt?

**Pull request merged**